cordova-lib/src/cordova/plugin.js

```diff
@@ -507,3 +527,117 @@ function versionString(version) {
+            return determinePluginVersionToFetch(
+                pluginInfo.versions,
+                pluginInfo.engines.cordovaDependencies,
+                pluginMap,
+                platformVersions,
+                cordovaVersion);
+        });
+    } else {
+        // If we have no engine, we want to fall back to the default behavior
+        return Q.fcall(function() {
+            return null;
+        });
+    }
+}
+
+function determinePluginVersionToFetch(allVersions, engine, pluginMap, platformMap, cordovaVersion) {
+    var upperBounds = [];
+    var versions = [];
+
+    for(var version in engine) {
+        if(version.indexOf('<') === 0 && semver.valid(version.substring(1))) {
+            // We only care about upper bounds that our project does not support
+            if(!satisfiesProjectRequirements(engine[version], pluginMap, platformMap, cordovaVersion)) {
+                upperBounds.push(version);
+            }
+        } else if(semver.valid(version) && allVersions.indexOf(version) !== -1) {
+            versions.push(version);
+        }
+    }
+
+    // Sort in descending order; we want to start at latest and work back
+    versions.sort(semver.rcompare);
+
+    for(var i = 0; i < versions.length; i++) {
+        for(var j = 0; j < upperBounds.length; j++) {
+            if(semver.satisfies(versions[i], upperBounds[j])) {
+                // Because we sorted in desc. order, if we find an upper bound
+                // that applies to this version (and the ones below) we can just
+                // quit (we already filtered out ones that our project supports)
+                return null;
+            }
+        }
+
+        // Check the version constraint
+        if(satisfiesProjectRequirements(engine[versions[i]], pluginMap, platformMap, cordovaVersion)) {
+            // Because we sorted in descending order, we can stop searching once
+            // we hit a satisfied constraint
+            var index = versions.indexOf(versions[i]);
+            if(index > 0) {
+                // Return the highest plugin version below the next highest
+                // constrainted version
+                var nextHighest = versions[index - 1];
+                return allVersions[allVersions.indexOf(nextHighest) - 1];
+            } else {
+                // Return the latest plugin version
+                return allVersions[allVersions.length - 1];
+            }
+        }
+    }
+
+    // No constraints were satisfied
+    return null;
+}
+
+function satisfiesProjectRequirements(reqs, pluginMap, platformMap, cordovaVersion) {
+    for (var req in reqs) {
+        var okay = true;
+
+        if(pluginMap[req]) {
+            okay = semver.satisfies(pluginMap[req], reqs[req]);
+        } else if(req === 'cordova') {
+            okay = semver.satisfies(cordovaVersion, reqs[req]);
+        } else if(req.indexOf('cordova-') === 0) {
+            // Might be a platform constraint
```

Comment:
I originally had it that way, but decided on requiring the `cordova-` prefix to avoid ambiguity with the actual platforms. That is, Android OS version versus cordova-android version.